server/src/main/java/com/epam/aidial/core/server/service/schemarichapps/ApplicationResourcesReplacementService.java

```
@@ -0,0 +1,114 @@
+    private final EncryptionService encryptionService;
+    private final ResourceService resourceService;
+
+    public ApplicationResourcesReplacementService(ConfigStore configStore, EncryptionService encryptionService, ResourceService resourceService) {
+        this.configStore = configStore;
+        this.encryptionService = encryptionService;
+        this.resourceService = resourceService;
+    }
+
+    public void replaceSchemaRichAppOwnResources(Application application, String targetApplicationUrl, Map<String, String> replacementLinks) {
+        if (application.getApplicationTypeSchemaId() == null) {
+            return;
+        }
+        String targetApplicationResourceFolderUrl = getTargetFolderForCustomAppFiles(targetApplicationUrl, encryptionService);
+        replacementLinks = extractApplicationOwnResourcesMapping(application, targetApplicationResourceFolderUrl, replacementLinks);
+        applyApplicationOwnResourcesMapping(application, replacementLinks);
+    }
+
+    private Map<String, String> extractApplicationOwnResourcesMapping(Application application, String targetApplicationResourceFolderUrl, Map<String, String> replacementLinks) {
+        List<ResourceDescriptor> applicationOwnResources = ApplicationTypeSchemaUtils.getFiles(
+                configStore.get(), application, encryptionService, resourceService);
+
+        Map<String, String> resultMapping = new HashMap<>();
+
+        for (ResourceDescriptor resource : applicationOwnResources) {
+            String resourceUrl = resource.getUrl();
+            if (resource.isFolder()) {
+                String folderReplacement = replacementLinks.entrySet().stream()
+                        .filter(entry -> entry.getKey().startsWith(resourceUrl))
+                        .map(Map.Entry::getValue)
+                        .findFirst()
+                        .orElseThrow(() -> new IllegalStateException("Missing replacement link for folder: " + resourceUrl));
```

Is it expected or just assertion? If expected, please throw IllegalArgumentException

---

I need to return an error to the user, assuming that the publication request was modified before approval and the service cannot satisfy the application dependencies (which were initially specified but have been changed or removed). It is becoming technically possible, as we aim to enable publication editions.

---

Speaking of the conditions, it is more about the IllegalStateException, as it is not about the argument (publication request id to approve) but rather about the state (publication request content).

---

Please check the java doc and throw IllegalArgumentException.